common/src/main/java/org/conscrypt/ConscryptEngineSocket.java

```diff
@@ -600,10 +612,18 @@ private void writeToSocket() throws IOException {
         private final ByteBuffer fromEngine;
         private final ByteBuffer fromSocket;
         private final int fromSocketArrayOffset;
+        private final AllocatedBuffer allocatedBuffer;
         private InputStream socketInputStream;
 
         SSLInputStream() {
-            fromEngine = ByteBuffer.allocateDirect(engine.getSession().getApplicationBufferSize());
+            if (bufferAllocator != null) {
+                allocatedBuffer = bufferAllocator.allocateDirectBuffer(
+                        engine.getSession().getApplicationBufferSize());
+                fromEngine = allocatedBuffer.nioBuffer();
+            } else {
+                allocatedBuffer = null;
+                fromEngine = ByteBuffer.allocateDirect(engine.getSession().getApplicationBufferSize());
+            }
             // Initially fromEngine.remaining() == 0.
             fromEngine.flip();
             fromSocket = ByteBuffer.allocate(engine.getSession().getPacketBufferSize());
```

Comment:
Is there any reason we would not want to use direct buffers here (and SSLOutputStream) if we have an allocator? It looks like many paths eventually use the buffer allocator later on to swap heap buffers into temprary direct buffers for native operations later, which we could avoid. I may be mistaken though.

Reply:
We pass the underlying array of the `fromSocket` buffer to `InputStream.read()`, so we need a non-direct buffer so that it has an array.  (I think we've measured that the benefits from being able to do that outweigh the costs of using a non-direct buffer elsewhere, but it's possible that's not correct/no longer correct.)